explorer/ast/ast_rtti.h

```diff
@@ -0,0 +1,143 @@
+  CARBON_DEFINE_RAW_ENUM_CLASS(C##Kind, int) {                              \
+    CARBON_##C##_KINDS(IGNORE, ENUMERATOR)                                  \
+  };                                                                        \
+  template <typename Derived>                                               \
+  class C##KindTemplate : public CARBON_ENUM_BASE_CRTP(C##Kind, Derived) {  \
+   private:                                                                 \
+    using Base = CARBON_ENUM_BASE_CRTP(C##Kind, Derived);                   \
+    friend class AstRttiNodeKind;                                           \
+                                                                            \
+   public:                                                                  \
+    using IsCarbonAstRttiNodeKind = void;                                   \
+                                                                            \
+    C##KindTemplate() = default;                                            \
+                                                                            \
+    /* This type can be explicitly converted from the generic node kind. */ \
+    explicit C##KindTemplate(AstRttiNodeKind base_kind)                     \
+        : Base(Base::FromInt(base_kind.AsInt())) {}                         \
+                                                                            \
+    CARBON_##C##_KINDS(IGNORE, CARBON_INLINE_ENUM_CONSTANT_DEFINITION)      \
+  };                                                                        \
+  class C##Kind : public C##KindTemplate<C##Kind> {                         \
+    using C##KindTemplate<C##Kind>::C##KindTemplate;                        \
+  };
+#define ENUMERATOR(E) E = AstRttiNodeKind::E.AsInt(),
+CARBON_AST_RTTI_KINDS(DEFINE_ENUM, IGNORE)
+#undef DEFINE_ENUM
+#undef ENUMERATOR
+
+// Helper to compute the number of derived classes for a class by repeatedly
+// adding one.
+#define ADD_ONE(C) +1
+#define NUMBER_OF_DERIVED_CLASSES(C) CARBON_##C##_KINDS(IGNORE, ADD_ONE)
+
+// Helper to find the first derived class of a class.
+#define EXPAND(A) A
+#define LPAREN (
+#define RPAREN )
+#define FIRST(A, ...) A
+#define COMMA_AFTER(A) A,
+#define FIRST_DERIVED_CLASS_IMPL(C) \
+  FIRST LPAREN CARBON_##C##_KINDS(IGNORE, COMMA_AFTER) RPAREN
+#define FIRST_DERIVED_CLASS(C) EXPAND(FIRST_DERIVED_CLASS_IMPL(C))
+
+// Helper to find the last derived class of a class. This works by converting
+// the list of derived classes A B C into
+//   APPLY(, DISCARD) APPLY(A, DISCARD) APPLY(B, DISCARD) APPLY(C, RETURN)
+// which then expands to
+//   DISCARD() DISCARD(A) DISCARD(B) RETURN(C)
+// then to
+//   C
+#define DISCARD(...)
+#define RETURN(A) A
+#define APPLY(E, M) M(E)
+#define LAST_HELPER(E) , DISCARD RPAREN APPLY LPAREN E
+#define LAST_DERIVED_CLASS_IMPL(C) \
+  APPLY LPAREN CARBON_##C##_KINDS(IGNORE, LAST_HELPER), RETURN RPAREN
```

Comment:
FWIW, I find this pretty inscrutable. I would likely have to just read the post-preprocessor text to understand it. =[

I'm less concerned about the complexity in things like enum_base.h where it is factored away, but here it seems like it'll be fairly front-and-center for folks trying to understand how the AST works.

Reply:
Hopefully people aren't going to look at this much -- I don't think people looked at these files much when they were generated rather than being full of macros. But I think this is a real cost of the macro approach compared to generating code. If we want the code to be readable (and eg visible in a debugger and in code navigation / xref tools), the code generation option seems better.